src/cmd/singularity/cli/actions.go

```diff
@@ -466,7 +464,56 @@ func execStarter(cobraCmd *cobra.Command, image string, args []string, name stri
 		}
 	}
 
-	if err := exec.Pipe(starter, []string{procname}, Env, configData); err != nil {
-		sylog.Fatalf("%s", err)
+	if engineConfig.GetInstance() {
+		stdout, stderr, err := instance.SetLogFile(name)
+		if err != nil {
+			sylog.Fatalf("failed to create instance log files: %s", err)
+		}
+
+		pipefd, err := exec.SetPipe(configData)
+		if err != nil {
+			sylog.Fatalf("failed to set pipe communication: %s", err)
+		}
+
+		cmd := osexec.Command(starter, "")
```

Comment:
Maybe rebase on master, then check out the new function `src/pkg/util/exec/pipe.go:PipeCommand()`, and see if that fits your use case here

Reply:
Unfortunately that won't fit here, need to redirect stdout/stderr for command file logging

Reply:
Hmm, I think it will still work here. Look at function signature of `PipeCommand()`:

`func PipeCommand(command string, args []string, env []string, data []byte) (*exec.Cmd, error)`

`PipeCommand()` returns `*exec.Cmd` type, which is same type as `cmd` you're using here. So I think you can do this:

```
cmd, err := exec.PipeCommand(starter, "", Env, configData)
if err != nil {
    //fatal
}

// do stderr.Seek...

cmd.Stdout = stdout
cmd.Stderr = stderr

cmdErr := cmd.Run()
```

Reply:
`PipeCommand()` func will automatically set environment variable `PIPE_EXEC_FD` to the fd of the pipe file for the `exec.Cmd` type, and will also ensure that the pipe file remains open and passed to `starter` binary.

Reply:
Ok I look too quickly, that would work. Thanks Michael !